libs/language-server/src/lib/validation/validation-registry.ts

```diff
@@ -18,25 +24,41 @@ import { validatePropertyBody } from './checks/property-body';
 import { validateRangeLiteral } from './checks/range-literal';
 import { validateRegexLiteral } from './checks/regex-literal';
 import { validateValuetypeDefinition } from './checks/valuetype-definition';
+import { ValidationContext } from './validation-context';
 
 /**
  * Registry for validation checks.
  */
 export class JayveeValidationRegistry extends ValidationRegistry {
   constructor(services: JayveeServices) {
     super(services);
+
     this.register<JayveeAstType>({
-      BlockDefinition: validateBlockDefinition,
-      ColumnLiteral: validateColumnLiteral,
-      ConstraintDefinition: validateConstraintDefinition,
-      BooleanExpression: validateExpression,
-      JayveeModel: validateJayveeModel,
-      PipeDefinition: validatePipeDefinition,
-      PipelineDefinition: validatePipelineDefinition,
-      PropertyBody: validatePropertyBody,
-      RangeLiteral: validateRangeLiteral,
-      RegexLiteral: validateRegexLiteral,
-      ValuetypeDefinition: validateValuetypeDefinition,
+      BlockDefinition: wrapCheck(validateBlockDefinition),
+      ColumnLiteral: wrapCheck(validateColumnLiteral),
+      ConstraintDefinition: wrapCheck(validateConstraintDefinition),
+      BooleanExpression: wrapCheck(validateExpression),
+      JayveeModel: wrapCheck(validateJayveeModel),
+      PipeDefinition: wrapCheck(validatePipeDefinition),
+      PipelineDefinition: wrapCheck(validatePipelineDefinition),
+      PropertyBody: wrapCheck(validatePropertyBody),
+      RangeLiteral: wrapCheck(validateRangeLiteral),
+      RegexLiteral: wrapCheck(validateRegexLiteral),
+      ValuetypeDefinition: wrapCheck(validateValuetypeDefinition),
     });
   }
 }
+
+type JayveeValidationCheck<T extends AstNode = AstNode> = (
+  node: T,
+  context: ValidationContext,
+) => MaybePromise<void>;
+
+function wrapCheck<T extends AstNode = AstNode>(
+  check: JayveeValidationCheck<T>,
+): ValidationCheck<T> {
+  return (node: T, accept: ValidationAcceptor): MaybePromise<void> => {
+    const context = new ValidationContext(accept);
+    return check(node, context);
+  };
+}
```

Comment:
This design says that each validation registered is independent from each other, right? Just wanna be sure :)

Reply:
Yes, now there is at most one validation function for each kind of AST node. That single validation function can now perform checks as needed and potentially abort.

The `warpCheck` function you highlighted here is used to initialize the `ValidationContext`, so each validation function gets a "fresh" context.